backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/ModularContractService.java

```diff
@@ -228,11 +252,84 @@ public void afterInvoiceReverse(@NonNull final I_C_Invoice invoiceRecord, @NonNu
 		{
 			return;
 		}
-		final FlatrateTermId flatrateTermId = Check.assumePresent(flatrateBL.getIdByInvoiceId(invoiceId), "FlatrateTermId should be present");
+		final FlatrateTermId flatrateTermId = getFlatrateTermIdByInvoiceId(invoiceId);
 
 		modularContractSettingsService.getById(modularContractSettingsId).getModuleConfigs().stream()
 				.filter(config -> config.isMatchingAnyOf(AVERAGE_CONTRACT_SPECIFIC_PRICE_METHODS))
 				.forEach(config -> modularContractLogService.updateAverageContractSpecificPrice(config, flatrateTermId, logHandlerRegistry));
+
+		if(invoiceBL.isFinalInvoiceOrFinalCreditMemo(invoiceRecord))
+		{
+			updateIsFinalInvoiced(flatrateTermId, false);
+		}
+	}
+
+	public FlatrateTermId getFlatrateTermIdByInvoiceId(@NonNull final InvoiceId invoiceId)
+	{
+		return Check.assumePresent(flatrateBL.getIdByInvoiceId(invoiceId), "FlatrateTermId should be present");
 	}
 
+	public void updateIsFinalInvoiced(@NonNull final InvoiceId invoiceId, final boolean isFinalInvoiced)
+	{
+		if(invoiceBL.isFinalInvoiceOrFinalCreditMemo(invoiceId))
+		{
+			updateIsFinalInvoiced(getFlatrateTermIdByInvoiceId(invoiceId), isFinalInvoiced);
+		}
+	}
+
+	public void updateIsFinalInvoiced(@NonNull final FlatrateTermId flatrateTermId, final boolean isFinalInvoiced)
+	{
+		final I_C_Flatrate_Term flatrateTerm = flatrateBL.getById(flatrateTermId);
+		flatrateTerm.setIsFinalInvoiced(isFinalInvoiced);
+		flatrateBL.save(flatrateTerm);
+	}
+
+	public void setPurchaseModularContractIdsIfExists(@NonNull final I_C_Order orderRecord, final boolean isErrorIfMoreThanOneFound)
+	{
+		if(!orderRecord.isSOTrx())
+		{
+			return;
+		}
+		for(final I_C_OrderLine orderLineRecord : orderDAO.retrieveOrderLines(orderRecord))
+		{
+			updatePurchaseModularContractId(orderLineRecord, isErrorIfMoreThanOneFound);
+			orderDAO.save(orderLineRecord);
+		}
+	}
+
+	public void updatePurchaseModularContractId(
+			@NonNull final I_C_OrderLine orderLineRecord,
+			final boolean isErrorIfMoreThanOneFound)
+	{
+		final OrderId orderId = OrderId.ofRepoId(orderLineRecord.getC_Order_ID());
+		final ProductId productId = ProductId.ofRepoId(orderLineRecord.getM_Product_ID());
+		final Set<FlatrateTermId> contractIds = contractProvider.getInitialPurchaseModularContractCandidatesForSalesOrderLine(orderId, productId);
+		final FlatrateTermId currentFlatrateTermId = FlatrateTermId.ofRepoIdOrNull(orderLineRecord.getPurchase_Modular_Flatrate_Term_ID());
+
+		if (contractIds.isEmpty())
+		{
+			orderLineRecord.setPurchase_Modular_Flatrate_Term_ID(-1);
```

Comment:
This could delete a contract form an order line that is not related with modular contracts! 
We need a way to make sure we only apply this logic to relevant lines.

Reply:
this is a new column for modular contract only